# Social Media Sanctions and Misinformation Sharing

There has been a lot of discussions on politics and sanctions in social media. Specifically, people with conservative beliefs argue that they are more likely to be sanctioned on social media than those with more liberal beliefs.

If this happens, however, that may not necessarily be the result of different policies on behalf of the social media companies. It may simply be that people with conservative beliefs post online more questionable content, and so they are more likely to be sanctioned even with a completely neutral sanction policy.

In a [recent analysis in Nature](https://www.nature.com/articles/s41586-024-07942-8), Mohshen Moshel et al. investigated whether social media sanctions are politically biased or they are the result of differences in misinformation sharing between people of different political beliefs. The authors argue that that media sanctions are not politically biased. Social media users with conservative political beliefs are more likely to post questionable or misinformation material, and so they are more likely to be sanctioned. In this assignment, you will follow through some of Moshel et al.'s research.

You will use their data, available at [mosleh_et_al_data.csv](mosleh_et_al_data.csv).

In [45]:
import pandas as pd
from scipy.stats import chi2_contingency

In [46]:
data = pd.read_csv('mosleh_et_al_data.csv')
data.head()

politics_sites2  politics_sites1  politics_followed  lowqual_pr2019_crowd  \
0        -0.231902        -1.168745          -0.080295              0.598883   
1         0.405261         1.060905           2.296375              0.774873   
2        -0.171067        -0.500076          -2.312730              0.548727   
3        -0.085814        -0.748035          -1.507481              0.611501   
4         0.595367         1.347730           2.010573              0.700217   

   lowqual_pr2019_fc  lowqual_afm  lowqual_mbfc  lowqual_lasser2019  \
0           0.279221     0.185315      0.292807            0.350000   
1           0.848639     0.614156      0.513147            0.975087   
2           0.242347     0.130950      0.242445            0.300000   
3           0.360119     0.180055      0.277793            0.316038   
4           0.698661     0.430655      0.554609            0.763889   

   suspended  politics_hashtag  ...  valence_neg  valence_neu  valence_pos  \
0          0                 0  ...     0.109935     0.813230     0.076835   
1          0                 1  ...     0.115985     0.793715     0.090250   
2          0                 0  ...     0.085765     0.817340     0.096915   
3          0                 0  ...     0.108680     0.781135     0.110190   
4          0                 1  ...     0.065145     0.860550     0.074320   

   politics_sites2_ideo_std  hashtag_ideo_std  barbera_std  \
0                  0.274317               0.0     0.909460   
1                  0.351800               0.0     0.157056   
2                  0.279591               0.0     0.445349   
3                  0.290277               0.0     0.256665   
4                  0.300823               0.0     0.431149   

   politics_sites1_std  repeated_tweet_count  freq_pr_1h  bad_domain_count  
0                   69                  86.0          30                 7  
1                  623                 133.0          26                20  
2                   32                   3.0           3                 6  
3                   55                 104.0          29                22  
4                   37                  36.0           8                 8  

[5 rows x 42 columns]

In [29]:
list(data.columns)

['politics_sites2',
 'politics_sites1',
 'politics_followed',
 'lowqual_pr2019_crowd',
 'lowqual_pr2019_fc',
 'lowqual_afm',
 'lowqual_mbfc',
 'lowqual_lasser2019',
 'suspended',
 'politics_hashtag',
 'lowqual_pr2019_crowdrep',
 'num_followers',
 'num_friends',
 'tweets_in_2wk',
 'botsentinel_score',
 'rewire_abuse',
 'rewire_hate',
 'rewire_profanity',
 'rewire_violent',
 'rewire_sexually_explicit',
 'jigsaw_toxicity',
 'jigsaw_severe_toxicity',
 'jigsaw_idenity_attack',
 'jigsaw_insult',
 'jigsaw_profanity',
 'jigsaw_threat',
 'rocklage_valence_std',
 'rocklage_extremity_std',
 'rocklage_emotionality_std',
 'rocklage_certainty_std',
 'liwc_political',
 'moral_outrage',
 'valence_neg',
 'valence_neu',
 'valence_pos',
 'politics_sites2_ideo_std',
 'hashtag_ideo_std',
 'barbera_std',
 'politics_sites1_std',
 'repeated_tweet_count',
 'freq_pr_1h',
 'bad_domain_count']

## Data Cleaning

In [44]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9000 entries, 0 to 8999
Data columns (total 42 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   politics_sites2            9000 non-null   float64
 1   politics_sites1            9000 non-null   float64
 2   politics_followed          7965 non-null   float64
 3   lowqual_pr2019_crowd       8945 non-null   float64
 4   lowqual_pr2019_fc          8945 non-null   float64
 5   lowqual_afm                8998 non-null   float64
 6   lowqual_mbfc               8999 non-null   float64
 7   lowqual_lasser2019         8999 non-null   float64
 8   suspended                  9000 non-null   int64  
 9   politics_hashtag           9000 non-null   int64  
 10  lowqual_pr2019_crowdrep    8945 non-null   float64
 11  num_followers              9000 non-null   int64  
 12  num_friends                9000 non-null   int64  
 13  tweets_in_2wk              9000 non-null   int64

In [13]:
data.loc[:, data.isna().any()]

politics_followed  lowqual_pr2019_crowd  lowqual_pr2019_fc  lowqual_afm  \
0             -0.080295              0.598883           0.279221     0.185315   
1              2.296375              0.774873           0.848639     0.614156   
2             -2.312730              0.548727           0.242347     0.130950   
3             -1.507481              0.611501           0.360119     0.180055   
4              2.010573              0.700217           0.698661     0.430655   
...                 ...                   ...                ...          ...   
8995                NaN              0.559728           0.257015     0.167228   
8996           1.689321              0.703698           0.720918     0.399594   
8997                NaN              0.627648           0.400227     0.217915   
8998           1.525494              0.764551           0.808780     0.388632   
8999          -1.411865              0.581819           0.252232     0.185156   

      lowqual_mbfc  lowqual_lasser2019  lowqual_pr2019_crowdrep  \
0         0.292807            0.350000                 0.700800   
1         0.513147            0.975087                 0.755118   
2         0.242445            0.300000                 0.658118   
3         0.277793            0.316038                 0.699165   
4         0.554609            0.763889                 0.725429   
...            ...                 ...                      ...   
8995      0.473955            0.336758                 0.673878   
8996      0.568832            0.772321                 0.727482   
8997      0.303897            0.355212                 0.714133   
8998      0.462972            0.863636                 0.732447   
8999      0.284862            0.318627                 0.686084   

      botsentinel_score  rewire_abuse  rewire_hate  ...  \
0                  15.0      0.180008     0.109561  ...   
1                  52.0      0.085648     0.133644  ...   
2                  17.0      0.156415     0.100484  ...   
3                  24.0      0.206408     0.079237  ...   
4                  56.0      0.084418     0.078898  ...   
...                 ...           ...          ...  ...   
8995               34.0      0.181040     0.084420  ...   
8996               50.0      0.100079     0.080027  ...   
8997               23.0      0.205299     0.097407  ...   
8998               57.0      0.250229     0.084796  ...   
8999               17.0      0.114230     0.081565  ...   

      rocklage_extremity_std  rocklage_emotionality_std  \
0                   0.445273                   1.081364   
1                   0.191000                   0.382000   
2                   0.716100                   1.634400   
3                   0.703071                   1.374143   
4                   0.382444                   0.888600   
...                      ...                        ...   
8995                0.647733                   1.196533   
8996                0.626571                   0.934429   
8997                0.831250                   1.517750   
8998                0.632300                   0.911300   
8999                0.532077                   1.083154   

      rocklage_certainty_std  liwc_political  moral_outrage  valence_neg  \
0                   1.176421           0.465       0.219615     0.109935   
1                   0.652917           0.680       0.183029     0.115985   
2                   1.241714           0.450       0.254935     0.085765   
3                   0.732737           0.410       0.229339     0.108680   
4                   1.241450           0.580       0.260533     0.065145   
...                      ...             ...            ...          ...   
8995                1.003000           0.520       0.207269     0.087090   
8996                1.180278           0.445       0.208994     0.067430   
8997                0.954929           0.355       0.272978     0.083680   
8998                1.100783           0.485      

This dataset contains columns representing both categorical and numerical data. Missing values are handled based on their specific type and significance in statistical analysis.

1. **Categorical Columns**
* `politics_followed`, `lowqual_pr2019_crowd`, `lowqual_pr2019_fc`, `lowqual_afm`, `lowqual_mbfc`, `lowqual_lasser2019`, `lowqual_pr2019_crowdrep`

We will replace missing values with the **most frequent value (mode)**.

2. ****

## Twitter Suspensions after the 2020 Election

Show, using crosstabulation, that accounts that shared `#Trump2020` during the election were about 4.4 times more likely to be subsequently suspended than the accounts that shared `#VoteBidenHarris2020` (relevant columns: `politics_hashtag` and `suspended`). Then perform a $\chi^2$ test on the contingency table and explain the results.

In [36]:
table = pd.crosstab(
  index=data['politics_hashtag'],
  columns=data['suspended'],
)
table

suspended            0    1
politics_hashtag           
0                 4295  200
1                 3623  882

Using the results of the Crosstab Table, we can see that the republicans were about 4.41 times more likely to be subsequently suspended than the accounts of the democrats.

In [43]:
ratio = table.loc[1, 1] / table.loc[0, 1]
ratio

4.41

By performing a $\chi^2$ test we get the following results:
  * $\chi^2$ statistic:
    * **Value:** 485.47
    * A high chi-square value, like 485.47, indicates a substantial deviation from what we would expect if there were no realtionship between account suspendion and political party.
  * p-Value:
    * **Value:** $1.38 \times 10^{-107}$
    * The p-value indicates the probability of observing a chi-square a chi-square value as extreme as 485.47 (or more extreme) under the null hypothesis that account suspension and political party affiliation are independent.
    * A p-value this small (essentially 0) is far below any conventional significance level. This strongly suggests that the null hypothesis of independence can be rejected.
  * Significance:
    * **Result:** True
    * The significance result confirms that the p-value is below the pre-determined significance threshold (**here 5%**).
We can observe from the results that the account suspension is significantly associated with the political party affiliation.

In [50]:
chi2, p, dof, expected = chi2_contingency(table)
significant = p < 0.05 # 5% significance level

chi2, p, significant

(485.46706126582285, 1.3808204811294585e-107, True)

## Distributions of Relative Frequency of Low Quality

Plot the distributions of low-quality sharing as determined by eight professional fact-checkers (column `lowqual_pr2019_fc`), grouped by `politics_hashtag`. The $x$-axis should be standardized using z-scores. The distributions should look like the following figure.

<img src="fig_1.svg"/>

Having done, that, proceed to do the same with low-quality sharing as determined by crowdsourcing from 971 participants from the USA (column `lowqual_pr2019_crowd`). The distributions should look like the following figure.

<img src="fig_2.svg" />

Show the results of the t-statistic for these two groups as well.

## Low-quality News Sharing on Twitter

We will proceed to examine low-quality news sharing on Twitter on a more quantitave manner, using the t-test. In particular, we will use the t-test based on groups of users depending on their `politics_hashtag` and various assessments of low-quality news sharing. We will quantify the effect size of the t-test using Cohen's $d$ and Hedges' $g$.

Cohen's $d$ is defined as the difference between two means divided by a standard deviation for the data, i.e.:

$$ d = \frac{\bar{x}_{1} - \bar{x}_{2}}{s} $$

$s$, the pooled standard deviation for two independent samples is defined as:

$$ s=\sqrt{\frac{(n_{1}-1)s_{1}^{2}+(n_{2}-1)s_{2}^{2}}{n_{1}+n_{2}-2}} $$

where $n_1$, $n_2$ is the size of each sample and the variance of $s_1$ is defined as:

$$ s_{1}^{2}=\frac{1}{n_{1}-1} \sum _{i=1}^{n_{1}}(x_{1,i}-{\bar {x}}_{1})^{2} $$

with the variance of $s_2$ defined similarly. The values of Cohen's $d$ can be interpreted as follows:

| Effect Size | $d$  |
|-------------|------|
| Very small  | 0.01 |
| Small	      | 0.20 |
| Medium	  | 0.50 |
| Large	      | 0.80 |
| Very large  | 1.20 |
| Huge        | 2.00 |

Hedges' $g$ corrects Cohen's $d$ for bias in small sample sizes and is defined as:

$$ 1 - \bigg(\frac{3}{4  (n_1 + n_2) - 9)}\bigg)d $$

You will perform a t-test statistic and report Cohen's $d$ and Hedges' $g$ for the following, grouped by `politics_hashtag`.

* `lowqual_pr2019_fc` (sites rated by fact-checkers)
* `lowqual_afm` (sites rated by Ad Fontes media, <https://adfontesmedia.com/>)
* `lowqual_mbfc` (sites rated by Media Bias/Fact Check, <https://mediabiasfactcheck.com/>)
* `lowqual_lasser2019` (sites rated by Laser et al., https://doi.org/10.1093/pnasnexus/pgac186)
* `lowqual_pr2019_crowd` (sites rated by crowdsourcing 971 participants from the USA)
* `lowqual_pr2019_crowdrep` (sites rated by the republicans among the 971 participants of above)

### Sharing Lower-quality News Sources and Conservative Ideology

To see if there is a correlation between lower-quality news sources and conservative ideology, you will perform pairwise correlations between, on the one hand:

 * `lowqual_pr2019_fc`
 * `lowqual_lasser2019`
 * `lowqual_mbfc`
 * `lowqual_afm`
 * `lowqual_pr2019_crowd`
 * `lowqual_pr2019_crowdrep`

and on the other hand:

 * `politics_followed` (political ideology estimated on the basis of Twitter accounts user followed)
 * `politics_hashtag`
 * `politics_sites1` (political ideology estimated on the basis of the news sites the users share, <https://doi.org/10.1126/science.aau2706>)
 * `politics_sites2` (political ideology estimated on the basis of the news sites ther users share, <https://doi.org/10.31219/osf.io/ch8gj>)

You should display the results in a heatmap like the following:

<img src="fig_3.svg"/>

## News Sharing Can Help Explain Suspension, Using Single Predictors

You will use the following predictors to predict whether a user was suspended or not:

  * `politics_hashtag`
  * `politics_sites1`
  * `politics_sites2`
  * `politics_followed`
  * `lowqual_afm`
  * `lowqual_mbfc`
  * `lowqual_lasser2019`
  * `lowqual_pr2019_fc`
  * `lowqual_pr2019_crowd`

The predictions will be with a probit model. The probit model is an alternative to logit, but while the latter uses the logit as the *link function*, the probit model uses the inverse of the cumulative distribution function (CDF) of the standard normal distribution (denoted as $\Phi^{-1}$) as the link function. Mathematically:

$$ \Phi^{-1}(P(Y = 1 | X)) = X\beta $$

where $P(Y = 1  | X)$ is the probability of the event occurring, $X$ represents the independent variables, and $\beta$ is the vector of coefficients.

The predictions will be evaluated with the AUC of the ROC curve. To get confidence intervals, repeat each prediction 100 times by using the boostrap method. 

Print the AUC and the confidence intervals in a table. Then use them to plot all your evaluations and the confidence intervals in a barchart like the following:

<img src="fig_4.svg" />

## News Sharing Can Help Explain Suspension, Using Multiple Predictors

You will also use multiple regression models to predict and explain suspension. 

Start by filing in missing values in your data by using the column mean. Then log transform, base 10, the variables that are related to Tweeter metrics, i.e.:

* `repeated_tweet_count` to `log_repeated_tweet_count`
* `freq_pr_1h` (maximum frequency of primary tweets in an hour) to `log_freq_pr_1h`
* `num_followers` to `log_num_followers`
* `num_friends` to `log_num_friends`

Introduce an additional variable, `log_fol`, which should the be base 10 logarithm of `num_friends` over `num_followers` ratio. To prevent out of domain errors, add 1 to each variable when using it in the transforms.

To reduce the effect of outliers, you will [winsorize](https://en.wikipedia.org/wiki/Winsorizing) the data. Winzorize from both sides at the top 1% the following variables:

* `valence_neg`
* `valence_neu`
* `valence_pos`
* `barbera_std`
* `politics_sites1_std`
* `politics_sites2_ideo_std`

The `valence_*`, variables were calculated using VADER (<https://doi.org/10.1609/icwsm.v8i1.14550>). The `barbera_std` variable was calculated based on the work of Barbera et al. (<https://doi.org/10.1177/0956797615594620>).

Winsorize the upper side of the following  variables at the top 1%:

* `moral_outrage`
* `rewire_abuse`
* `rewire_hate`
* `rewire_profanity`
* `rewire_violent`
* `rewire_sexually_explicit`
* `jigsaw_toxicity`
* `jigsaw_severe_toxicity`
* `jigsaw_idenity_attack`
* `jigsaw_insult`
* `jigsaw_profanity`
* `jigsaw_threat`

The `moral_outrage` variable was calculated based on the work of Brady et al. (<https://doi.org/doi:10.1126/sciadv.abe5641>). The `rewire_*` variables were calculated using the Rewire Online API (acquired by ActiveFence in 2023)., The `jigsaw_*` variables were calculated using the Google Jigsaw Perspective API (<https://jigsaw.google.com/>, <https://perspectiveapi.com/>).

You will derive and use a single PCA component, `pc1_misinfo`, for low quality news sharing, that is, the variables:

* `lowqual_pr2019_fc`
* `lowqual_afm`
* `lowqual_lasser2019`
* `lowqual_mbfc`

Show the explained variance ratio. Then project the low quality news sharing dimensions to this new dimension.

We will work in the same way for the political orientation, getting a single PCA component, `pc1_politics`, for the variables:

* `politics_hashtag`
* `politics_followed`
* `politics_sites1`
* `politics_sites2`

Show the explained variance ratio and project the political orientation dimensions to this new dimension.

Introduce a new variable, `extremity`, which is the absolute value of the PCA component you got.

Also do the same for `log_followers`, `log_friends`, `log_fol`, producing `pc1_fol`.

Regarding the harmful language variables, perform PCA for *three* components, `pc1_harmful_language`, `pc2_harmful_language`, `pc3_harmful_language`, on the winsorized data of:

* `rewire_abuse`
* `rewire_hate`
* `rewire_profanity`
* `rewire_violent`
* `rewire_sexually_explicit`
* `jigsaw_toxicity`
* `jigsaw_severe_toxicity`
* `jigsaw_idenity_attack`
* `jigsaw_insult`
* `jigsaw_profanity`
* `jigsaw_threat`

Why use three components instead of one in this case? Try to interpret each of the principal components.

Return to producing a single PCA component, `pc1_valence`, for the winsorized data of the valence variables:

* `valence_neg`
* `valence_neu`
* `valence_pos`

Before proceeding, normalize the independent variables by taking their z-scores. You will use the following z-scored independent variables (winsorized as stated above):

* `pc1_politics`
* `pc1_misinfo`
* `pc1_fol`,
* `pc1_harmful_language`
* `pc2_harmful_language`
* `pc3_harmful_language`
* `pc1_valence`
* `tweets_in_2wk`
* `botsentinel_score`
* `extremity`
* `w_moral_outrage`
* `liwc_political`
* `log_freq_pr_1h`
* `log_repeated_tweet_count`
* `barbera_std`
* `politics_sites1_std`
* `politics_sites2_ideo_std`

Carry out a probit regression to predict suspension.

Before going on to interpret your results, perform the [Bonferroni correction](https://en.wikipedia.org/wiki/Bonferroni_correction) and the [Holm-Bonferroni correction](https://en.wikipedia.org/wiki/Holm%E2%80%93Bonferroni_method) correction. The purpose of both of these tests is to reduce the likelihood of false positives (Type I errors) when conducting multiple statistical tests.

In the Bonferroni correction, a hypothesis is rejected with an ajustment to the significance level $\alpha$:

$$ p < \frac{\alpha}{m} $$

which means that the $p$ values are adjusted by:

$$ p' = p \times m $$

where $m$ is the number of tests.

The Bonferroni-Holm method is less conservative compared to Bonferroni. It sorts the $p$-values in ascending order and compares them to less stringent thresholds:

$$ p_i \le \frac{\alpha}{m - i + 1} $$ 

where $i$ is the $i$-th smallest $p$-value. Starting with the smallest $p$-value, hypotheses are rejected sequentially until a $p$-value fails to meet the threshold. That means that the $p$ values are adjusted by:

$$ p'_i = p_i \times (m - i + 1) $$

Holm-Bonferroni applies progressively smaller corrections as it moves through the sorted $p$-values, making it less conservative and less likely to fall into false negatives (Type II errors).

Based on the results, do you think that political orientation influences suspension?

Do the same analysis using logit regression.

## Submission Instructions

You must submit your assignment as a Jupyter notebook that will contain the full code and documentation of how you solved the questions, and all other necessary files. Your submission must be fully replicable: that is, somebody reading it must be able to do exactly what you did and obtain the same results.

The documentation must be at the level where somebody that has some knowledge of Python can understand exactly what you are doing and why. Your output must be as user-friendly as possible. That means that your output tables should not include zillions of columns that are not needed for your analysis.

## Honor Code

You understand that this is an individual assignment, and as such you must carry it out alone. You may seek help on the Internet, on ChatGPT/Gemini/etc., by Googling or searching in StackOverflow for general questions pertaining to the use of Python and pandas libraries and idioms. However, it is not right to ask direct questions that relate to the assignment and where people will actually solve your problem by answering them. You may discuss with your colleagues in order to better understand the questions, if they are not clear enough, but you should not ask them to share their answers with you, or to help you by giving specific advice.